js/web/lib/wasm/wasm-training-core-impl.ts

```diff
@@ -290,10 +285,84 @@ export const runTrainStep = async(
     if (wasm._OrtTrainingRunTrainStep) {
       const errorCode = wasm._OrtTrainingRunTrainStep(
           trainingSessionId, inputValuesOffset, inputCount, outputValuesOffset, outputCount, runOptionsHandle);
+      ifErrCodeCheckLastError(errorCode, 'failed to call OrtTrainingRunTrainStep in the WebAssembly layer');
+    } else {
+      throw new Error(NO_TRAIN_FUNCS_MSG);
+    }
 
-      if (errorCode !== 0) {
-        checkLastError('failed to call OrtTrainingRunTrainStep in the WebAssembly layer');
-      }
+    return moveOutputToTensorMetadataArr(outputValuesOffset, outputCount, outputTensorHandles, outputTensors);
+  } finally {
+    wasm.stackRestore(beforeRunStack);
+
+    inputTensorHandles.forEach(v => wasm._OrtReleaseTensor(v));
+    outputTensorHandles.forEach(v => wasm._OrtReleaseTensor(v));
+    inputOutputAllocs.forEach(p => wasm._free(p));
+
+    if (runOptionsHandle !== 0) {
+      wasm._OrtReleaseRunOptions(runOptionsHandle);
+    }
+    runOptionsAllocs.forEach(p => wasm._free(p));
+  }
+};
+
+export const runOptimizerStep =
+    async(trainingSessionId: number, options: InferenceSession.RunOptions): Promise<void> => {
```

Comment:
can you check with Yulong whether we need stackSave and restore calls here given we are not really passing any inputs or getting any output from wasm.

Reply:
Yep, Yulong confirmed that there is no need to call stack restore here since there are no stackAlloc calls